internal/configs/virtualserver_test.go

```diff
@@ -759,6 +764,74 @@ func TestGenerateUpstream(t *testing.T) {
 	}
 }
 
+func TestGenerateUpstreamWithKeepalive(t *testing.T) {
+	name := "test-upstream"
+	noKeepalive := 0
+	keepalive := 32
+	endpoints := []string{
+		"192.168.10.10:8080",
+	}
+	isPlus := false
+
+	tests := []struct {
+		upstream  conf_v1alpha1.Upstream
+		cfgParams *ConfigParams
+		expected  version2.Upstream
+	}{
+		{
+			conf_v1alpha1.Upstream{Keepalive: &keepalive},
+			&ConfigParams{
+				Keepalive: 21,
+			},
+			version2.Upstream{
+				Name: "test-upstream",
+				Servers: []version2.UpstreamServer{
+					{
+						Address: "192.168.10.10:8080",
+					},
+				},
+				Keepalive: 32,
+			},
+		},
+		{
+			conf_v1alpha1.Upstream{},
+			&ConfigParams{
+				Keepalive: 21,
+			},
+			version2.Upstream{
+				Name: "test-upstream",
+				Servers: []version2.UpstreamServer{
+					{
+						Address: "192.168.10.10:8080",
+					},
+				},
+				Keepalive: 21,
+			},
+		},
+		{
+			conf_v1alpha1.Upstream{Keepalive: &noKeepalive},
+			&ConfigParams{
+				Keepalive: 21,
+			},
+			version2.Upstream{
+				Name: "test-upstream",
+				Servers: []version2.UpstreamServer{
+					{
+						Address: "192.168.10.10:8080",
+					},
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		result := generateUpstream(name, test.upstream, endpoints, isPlus, test.cfgParams)
+		if !reflect.DeepEqual(result, test.expected) {
+			t.Errorf("generateUpstream() returned %v but expected %v", result, test.expected)
```

Comment:
please see the comment for the test `TestUpstreamHasKeepalive` about understanding which input causes a failure.